### PR TITLE
cmake,lib,base: Export obs-websocket-api as a target

### PR DESCRIPTION
**Pull request by tytan652**
### Description

Relies on:
- https://github.com/obsproject/obs-studio/pull/10100

This enables the installation of the header in the include directory by exporting it as an interface target

### Motivation and Context

No more copy-pasting the header to use it

### How Has This Been Tested?

Tested OS(s): 
- Arch Linux

Windows and macOS headers needs to be checked
- Tested on Windows,  `--component Development` adds the header

### Types of changes

- Other Enhancement (anything not applicable to what is listed)

### Checklist:

-  [x] I have read the [Contributing Guidelines](https://github.com/obsproject/obs-websocket/wiki/Contributing-Guidelines).
-  [x] All commit messages are properly formatted and commits squashed where appropriate.
-  [x] My code is not on `master` or a `release/*` branch.
-  [x] The code has been tested.
-  [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by tt2468** [comment]
Looks good to me, just curious if Pat has any comments on doing this

**Comment by tytan652** [comment]
This PR, the updated CMake of the obs-browser-api PR and the obs-studio PR that they rely on were based on a discussion I had with Pat about shipping those headets.

So for the 3 PRs, I'm waiting for Pat's feedback.

File: CMakeLists.txt
Comment (PatTheMav) [inline CMakeLists.txt:8]: Is there a specific reason you use `add_subdirectory` here? If you just wish to put the API stuff in a separate cmake file, you can use `include` (which also has the added benefit of not creating a new directory scope).

At this point in time it makes more sense to use an include and through the necessary code directly into the legacy script file, because there is no expectation of long-term maintainability of them (we'll probably remove the legacy stuff before that).

File: lib/CMakeLists.txt
Comment (PatTheMav) [inline lib/CMakeLists.txt:9]: This doesn't seem correct to me - the source list should just be `obs-websocket-api.h` in all situations - the path is resolved by `target_include_directories` below, where a differentiation between build interface and install interface is needed.

File: lib/CMakeLists.txt
Comment (PatTheMav) [inline lib/CMakeLists.txt:18]: The `FOLDER` property for the API as well as obs-websocket itself needs to be adjusted to say the same (`plugins/obs-websocket`) to group them.

File: lib/cmake/obs-websocket-apiConfig.cmake.in
Comment (PatTheMav) [inline lib/cmake/obs-websocket-apiConfig.cmake.in:5]: Is `libobs` the only dependency for the API header?

File: src/WebSocketApi.h
Comment (PatTheMav) [inline src/WebSocketApi.h:26]: Probably semantics, but removing the `obs.h` include here makes sense if the API implementation here does not use `libobs` API methods by itself.

The moment it directly does so, it needs to take care of the include itself and cannot rely that the websocket-API itself implicitly includes everything it needs.

Comment (PatTheMav) [reply src/WebSocketApi.h:26]: In practice the pre-processor might resolve this automatically, but that makes it harder for maintenance:

The `#include` directives at the top of a header also serve as documentation (as in: "These are the headers/libraries I implement functionality from").

File: lib/cmake/obs-websocket-apiConfig.cmake.in
Comment (tytan652) [reply lib/cmake/obs-websocket-apiConfig.cmake.in:5]: Yes, `proc_handler_t` is part of `libobs`.

File: CMakeLists.txt
Comment (tytan652) [reply CMakeLists.txt:8]: I prefer subdirectory "isolation" over `include` that requires to twist your mind about the file is not "executed" where it is but where it is included.

File: lib/CMakeLists.txt
Comment (tytan652) [reply lib/CMakeLists.txt:9]: I just followed those explanations:

https://crascit.com/2016/01/31/enhanced-source-file-handling-with-target_sources/

> The paths used for the project’s own build will be specific to the machine on which that build is performed and to the directory in which it is built. Once installed, those directories will likely not be accessible, so the paths would need to be replaced with something else that makes sense for the installed set of files. This is made possible using the BUILD_INTERFACE and INSTALL_INTERFACE generator expressions. The following example shows how to provide different paths for the same file in the two different contexts:
>
>```cmake
>target_sources(myHeaderOnly
>    PUBLIC
>        $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/algo.h>
>        $<INSTALL_INTERFACE:include/algo.h>
>)
>install(FILES algo.h DESTINATION include)
>```

Comment (PatTheMav) [reply lib/CMakeLists.txt:9]: Yeah that advice is for old CMake versions and isn't necessary anymore ever since CMake gained the ability to automatically treat sources added to a target as relative to the source directory (changed in CMake 3.13, released in 2018).

The only place where that differentiation is needed is when the sources added use an absolute path.

Otherwise just specifying a different include directory based on build or install interface is sufficient (see also how we handle public headers in `libobs`).

File: CMakeLists.txt
Comment (PatTheMav) [reply CMakeLists.txt:8]: In this case it just creates lots of overhead - `add_directory` creates a new directory scope with all the baggage that comes with it, but you only use the subdirectory to house CMake-specific files.

Instead the public library is a "feature" of obs-websockets itself and thus should be integrated into it. Using an include just makes maintenance easier as all CMake code specific to the public API is contained in there, but that's a distinction for us as maintainers but irrelevant to CMake.

Comment (tytan652) [reply CMakeLists.txt:8]: Also this:
```
set_target_properties(
  obs-websocket-api
  PROPERTIES
             …
             PUBLIC_HEADER lib/obs-websocket-api.h)
```
Does not inspire me that the header will be without the `lib` folder once installed.

Comment (PatTheMav) [reply CMakeLists.txt:8]: `PUBLIC_HEADER` does not support directory structures - all specified files will be flattened into a single directory (which is why we needed to add extra code to maintain `libobs` custom subdirectory structure).

Comment (tytan652) [reply CMakeLists.txt]: Did changes, please review them.

File: lib/CMakeLists.txt
Comment (tytan652) [reply lib/CMakeLists.txt]: I made big changes, so please re-review if those properties are correctly set.

**Review by PatTheMav — APPROVED** [review]

File: cmake/obs-websocket-api.cmake
Comment (PatTheMav) [inline cmake/obs-websocket-api.cmake:14]: The folder property is actually superfluous here because for INTERFACE libraries the sources are added directly into consumer's source list, they'll never appear as their own entries in MSVC' or Xcode's project view, but it also doesn't break anything.

**Pull request merged**
